### PR TITLE
fix: set subindices directly via access path

### PR DESCRIPTION
**Pull request by JohannesMessner**
Issue is described here: https://github.com/jina-ai/docarray/issues/489

### CODE REVIEW

File: docarray/array/storage/base/getsetdel.py [hunk 1]
```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
...
+        if isinstance(set_index, tuple):  # handled later in recursive call
+            return
+        if isinstance(set_index, str) and set_index.startswith('@'):
+            # 'nested' (non root-level) set, update entire subindex directly
+            _check_valid_values_nested_set(DocumentArray(self[set_index]), docs)
+            if set_index in subindices.keys():
+                subindex_da = subindices[set_index]
+                with subindex_da:
+                    subindices[set_index].clear()
+                    subindices[set_index].extend(docs)
```
Comment (alaeddine-13) [inline docarray/array/storage/base/getsetdel.py:202]: why not
```suggestion
                with subindex_da:
                    subindex_da.clear()
                    subindex_da.extend(docs)
```

File: docarray/array/storage/base/getsetdel.py [hunk 2]
```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
...
-                    da.extend(DocumentArray(docs)[selector])  # same hack here
+    def _update_subindices_set(self, set_index, docs):
+        subindices = getattr(self, '_subindices', None)
+        if not subindices:
+            return
+        if isinstance(set_index, tuple):  # handled later in recursive call
+            return
+        if isinstance(set_index, str) and set_index.startswith('@'):
+            # 'nested' (non root-level) set, update entire subindex directly
+            _check_valid_values_nested_set(DocumentArray(self[set_index]), docs)
```
Comment (alaeddine-13) [inline docarray/array/storage/base/getsetdel.py:197]: do we need to wrap into `DocumentArray` ? it's cast to DocumentArray inside the check function

File: docarray/array/storage/base/getsetdel.py [hunk 3]
```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
...
+            if set_index in subindices.keys():
```
Comment (alaeddine-13) [inline docarray/array/storage/base/getsetdel.py:198]:
```suggestion
            if set_index in subindices:
```

File: docarray/array/storage/base/getsetdel.py [hunk 4]
```diff
@@ -173,15 +186,26 @@ def _set_docs_by_ids(self, ids, docs: Iterable['Document'], mismatch_ids: Dict):
...
+                subindex_da = subindices[set_index]
+                with subindex_da:
+                    subindices[set_index].clear()
+                    subindices[set_index].extend(docs)
+        else:  # root level set, update subindices iteratively
+            for subindex_selector, subindex_da in subindices.items():
+                old_ids = DocumentArray(self[set_index])[subindex_selector, 'id']
+                with subindex_da:
+                    del subindex_da[old_ids]
+                    subindex_da.extend(DocumentArray(docs)[subindex_selector])
```
Comment (alaeddine-13) [inline docarray/array/storage/base/getsetdel.py:208]: probably this is not the correct way to update the subindex but I'm afraid there is no better efficient way:
if you delete document chunks (the root document is positioned in a middle position) from the subindex and you append its new chunks to the subindex, we are changing the order of docs within the subindex, and I think order matters in some minor cases

Comment (JohannesMessner) [reply docarray/array/storage/base/getsetdel.py:208]: This is correct, we knew about this and decided that we don't want to make guarantees about the order in the subindex. If you care about it i think you can still get the order from the serialized chunks that are stored in the root index

File: docarray/array/storage/base/getsetdel.py (hunk 2 above)
Comment (JohannesMessner) [reply docarray/array/storage/base/getsetdel.py:197]: You are right, I'll change it

Comment (JohannesMessner) [inline docarray/array/storage/base/getsetdel.py:197]:
```suggestion
            _check_valid_values_nested_set(self[set_index], docs)
```

**Comment by alaeddine-13** [comment]
I'm also wondering, shouldn't we have the same logic for _update_subindices_del ?

**Review by alaeddine-13 — APPROVED** [review]

**Pull request merged**